src/bls12_381/ec.rs

```diff
@@ -194,6 +260,34 @@ macro_rules! curve_impl {
                 (*self).into()
             }
 
+            fn hash(msg: &[u8]) -> Self {
+                // the key must be of at least 128 bits.
+                assert!(msg.len() >= 16);
+
+                // The construction of Foque et al. requires us to construct two
+                // "random oracles" in the field, encode their image with `sw_encode`,
+                // and finally add them.
+                // We construct them appending to the message the string
+                // $name_$oracle
+                // For instance, the first oracle in group G1 appends: "G1_0"
+                let mut hasher = Blake2b::new();
+                hasher.input(msg);
+                hasher.input($name.as_bytes());
```

Comment:
The hasher at this point can be cloned so that the state can be reused for the other oracle.